source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -998,6 +998,338 @@ See the `Load Balancer Specification <../load-balancers/load-balancers.rst#event
+     - Value
+
+   * - message
+     - String
+     - "Connection ready"
+
+   * - driverConnectionId
+     - Int
+     - The driver-generated ID for the connection as defined in `Connection <#connection>`_.
+
+The unstructured form SHOULD be as follows, using the values defined in the structured format above to fill in placeholders as appropriate:
+
+  Connection ready with driver-generated ID {{driverConnectionId}} to {{serverHost}}:{{serverPort}}
+
+Connection Closed Message
+-------------------------
+In addition to the common fields defined above, this message MUST contain the following key-value pairs:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Key
+     - Suggested Type
+     - Value
+
+   * - message
+     - String
+     - "Connection closed"
+
+   * - driverConnectionId
+     - Int
+     - The driver-generated ID for the connection as defined in `Connection <#connection>`_.
+
+   * - reason
+     - String
+     - A string describing the reason the connection was closed. The following strings MUST be used for each possible reason
+       as defined in `Events <#events>`_ above:
+
+       - Stale: "Connection became stale because the pool was cleared"
+       - Idle: "Connection has been available but unused for longer than the configured max idle time"
+       - Error: "An error occurred while using the connection"
+       - Pool closed: "Connection pool was closed" 
+
+The unstructured form SHOULD be as follows, using the values defined in the structured format above to fill in placeholders as appropriate:
+
+  Connection closed with driver-generated ID {{driverConnectionId}} to {{serverHost}}:{{serverPort}}. Reason: {{reason}}
+
+Connection Checkout Started Message
+-----------------------------------
+In addition to the common fields defined above, this message MUST contain the following key-value pairs:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Key
+     - Suggested Type
+     - Value
+
+   * - message
+     - String
+     - "Connection checkout started"
+
+The unstructured form SHOULD be as follows, using the values defined in the structured format above to fill in placeholders as appropriate:
+
+  Checkout started for connection to {{serverHost}}:{{serverPort}}
+
+Connection Checkout Failed Message
+-----------------------------------
+In addition to the common fields defined above, this message MUST contain the following key-value pairs:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 1 1
+
+   * - Key
+     - Suggested Type
+     - Value
+
+   * - message
+     - String
+     - "Connection checkout failed"
+
+   * - reason
+     - String
+     - A string describing the reason checkout. The following strings MUST be used for each possible reason
+       as defined in `Events <#events>`_ above:
+
+       - Timeout: "Wait queue timeout elapsed without a connection becoming available"
+       - ConnectionError: "An error occurred while trying to establish a new connection"
```

Comment:
It would probably be very useful if the actual error that caused the failure to occur were included here. That way the user could discern between a network issue, auth issue, OSCP responder problem, etc. I know the existing CMAP events are also deficient in this regard, but would this be a good opportunity to fix that, at least for logging messages?

Reply:
I agree that would be useful and I was thinking about it as I made these spec changes actually that it was funny we don't include the actual error, both here and for `ConnectionClosedEvent`. 
I'm happy to add the errors to the log messages, but a caveat worth mentioning first: for any driver implementing logging via a CMAP event listener, this is going to implicitly require adding the error to the corresponding monitoring events. I don't think that's a bad thing, but it would mean drivers would potentially end up with API that is not directly tested (it would at least be implicitly tested via logging though). we could file a follow up ticket around that though. alternatively, we could formally add it the errors to the events here too, but I think that the work to formally write tests around that (including possibly adding to the CMAP test format) would be too much scope creep. 
LMK if you think just adding the errors to the log messages and filing a follow up ticket around formally adding them to the events is reasonable.

Reply:
Adding it to just the log messages and deferring the event changes to a follow up ticket sgtm. I can file the ticket for it.

Reply:
I pushed changes to the spec and the tests for this. they pass in Rust. FYI @BorisDog - can you review and see if adding the errors to .NET log messages is possible and if the tests pass for you?

Reply:
Filed https://jira.mongodb.org/browse/DRIVERS-2495

Reply:
Sorry for the delay. Yes adding errors is straightforward and the tests pass.

Reply:
woo! thanks @BorisDog